_posts/2019-08-01-my-own-summer-neural-network.md

```diff
@@ -0,0 +1,321 @@
+
+I got pretty low frame rate of couple of images per second, but it turned out that this frame rate is actually enough for my simple Endless Flighter game. Note that images in the learning data are cropped so that the bars behind the player are not visible. This originates from my assumption, that the bars already passed do not matter anymore. When you play the game, you glaze your vision to the upper part of the screen.
+
+### 2. Capture Lots of Data by Flying Manually
+
+I wanted to make this work end-to-end with a minimal training data set, because I did not want to spent hours flying the most dull game in the history for hours and then see that it didn't work and the data is corrupt somehow. So I decided just to fly for a couple of minutes and try to achieve end-to-end solution which will probably be really bad because of the small amount of data.
+
+### 3. Balance the Data
+
+I learned, that overfitting is a common problem with neural networks. Overfitting means that you have too much of certain values in the learning data and then the network would pretty much always output saturated values. In this example, flying straight is way more common than steering, so overfitted network would always fly straight. To overcome this, you'll have to drop additional "flying straight" examples to match the steering examples. The goal is to get roughly the same amount of targets (keyboard states) to learning data.
+
+One of the most interesting things is that the ordering of the examples do not matter. Most of the people would intuitively think that playing this sort of game is contextual, meaning that previous examples affect the current status. But this is not how neural networks work, as neural network do not "see" the previous images. Learning data is shuffled, and replaying the balanced learning data visually does not make any sense for human anymore - it's just a collection of images and corresponding keyboard states in random order.
+
+I also made a small mistake in the learning data capturing stage, because I captured keyboard presses in a form of an array of `[a_pressed, s_pressed]`:
+
+```
+[0, 0] # nothing pressed
+[1, 0] # a pressed
+[0, 1] # s pressed
+```
+
+[TensorFlow](https://www.tensorflow.org/) produces *predictions* for the probability of the output for input, so you would get back an array of two elements. So for an screenshot, you would get back something like `[0.493, 0.984]` which would suggest to press the `s` button, because second value in the prediction array - which denominates the probability for `s` key pressed - is larger. Usually you select the action with highest probability and this would mean that our self playing agent would *always* turn to some direction, because there is no probability for going straight. To overcome this, I just simply mapped the data as following:
```

Comment:
```suggestion
[TensorFlow](https://www.tensorflow.org/) produces *predictions* for the probability of the output for input, so you would get back an array of two elements. So for a screenshot, you would get back something like `[0.493, 0.984]` which would suggest to press the `s` button, because the second value in the prediction array - which denominates the probability for `s` key pressed - is larger. Usually you would select the action with the highest probability and this would mean that our self playing agent would *always* turn to some direction, because there is no probability for going straight. To overcome this, I just simply mapped the data as following:
```